### ISSUE TITLE
Invalid call target address / packed exe (16-bit DOS)

### ISSUE THREAD

**Issue by movAX13h** [uxmal/reko#823]
Hello,

thank you for working on Reko. I'm delighted to see all the progress made recently. Especially for 16-bit DOS executables, there are not many good decompilers and tools available.

I have a small 16-bit DOS executable (abandonware, 43kb). Reko stops when reaching a far jump to an address outside of the range of the code segment that is not initialised (because the program does something before that makes the jump a valid jump). It's the first function called, so, the rest of the program escapes the decompiler.

All of this can be seen in this screenshot (using Reko v0.9):
![image](https://user-images.githubusercontent.com/1974959/71551271-75451c80-29e3-11ea-8927-4bc37f02e22d.png)

IDA-Pro is also having troubles with this exe (does not even get the one function Reko finds) so I think it might be a good point to figure out how to handle these cases so that at least some other functions of the binary get identified - even if results might be less reliable.

**Comment by movAX13h** [comment]
The binary is start.exe from [Transarctica](https://abandonwaregames.net/game/transarctica#download) (which btw. is the exact same exe that is used for the game Ishar 2).

**Comment by uxmal** [comment]
At first glance it appears like the binary has been packed with a packer Reko is not familiar with. It should be possible to write an unpacker script for it. I will be back in front of my development rig next week, at which point I can take a closer look.

**Comment by movAX13h** [comment]
Oh, you're right, thanks! I was able to unpack it with TRON for DOS. 
Sadly the unpacked binary crashes Reko.

**Comment by uxmal** [comment]
I would love to know specifically how Reko crashed after you unpacked the exe -- so I can fix the crash.

**Comment by movAX13h** [comment]
It's running into a System.StackOverflowException in Decompiler\Analysis\SsaTransform.IdentifierTransformer.cs, method ReadBlockLocalVariable()

See callstack bottom right: 
![image](https://user-images.githubusercontent.com/1974959/71559983-710d1380-2a64-11ea-9662-61dc5faddd18.png)

I'm trying to figure out how to compile the solution in VS (the instance in the screenshot is the build that came with the repo) ... the call to cmake.exe fails with error code 9009 (also visible in the screenshot). 

Find attached the unpacked exe. When unpacking the exe using Tron it prints "processed with: DIET V1.20".

[start_unpacked.zip](https://github.com/uxmal/reko/files/4008304/start_unpacked.zip)

**Comment by movAX13h** [comment]
There's a doc file (plain text) coming with Tron that contains a list of packers commonly used in DOS and other details. I've attached a copy of Tron just in case you don't have it (unregistered version).

[Tron (DOS exe unpacker)](https://github.com/uxmal/reko/files/4008341/tron.zip)

**Comment by uxmal** [comment]
Reactivating this to track the issue.

**Comment by smx-smx** [comment]
@movAX13h you need to install CMake (which should be in the C/C++ workload) in the Visual Studio Installer

**Comment by movAX13h** [comment]
@smx-smx thank you for responding.
tl;dr: solved by adding cmake bin dir to PATH env var

Might be an issue with VS. I've tried both, VS2017 and VS2019 (latest versions), installed cmake build support for both, rebooted, cleaned/rebuilt the solution ... it always complained about not finding cmake.exe in both versions of VS. I also tried to run cmake from a dev console launched from within VS and strangely there it was available. After having no success trying several steps and combinations, I decided to add the path to the bin dir of cmake to the PATH system environment variable, restarted VS, cleaned the solution and everything compiled successfully.

**Comment by uxmal** [comment]
Still working on this; fixing the stack overflow exposes other issues.

**Comment by uxmal** [comment]
The most recent commit seems to fix the worst problems. I couldn't resist the temptation to write my own unpacker using the executable's own unpacking code and Reko's support for OllyDbg scripting. If you're curious, the unpacker code is in `RekoUnpacker_0823.osc`, as follows:
```
var addrFarjmp
var addrUnpacker
var selector

// Run until just before the first long JMP.
bp cs:0035
run

// Execute the long jump.
sti

// Break at the instruction after relocation is done.
bp cs:064D

// Break at the far JMP into the original code.
bp cs:0666

relocation_loop:
run

cmp ip,0666
je relocation_done

// Read the relocation
mov selector,[es:bx],2
msg $"Relocation at {es:bx} = {selector}"
reko.addseg es:bx,selector
jmp relocation_loop

relocation_done:
// Single step into the original code.
sti
msg $"Entry point {cs:ip}"
dpe "dontcare",cs:ip
ret
```
